### PR TITLE
API error handling

### PR DESCRIPTION
**Pull request by gpatrac**
Added GlobalException handlers with @ControllerAdvice to handle api errors, SFA-412

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/api/accounts/CompanyAccountsApplication.java [hunk 1]
```diff
@@ -18,39 +18,40 @@
 @SpringBootApplication
 public class CompanyAccountsApplication implements WebMvcConfigurer {
 
-    @Autowired
-    private TransactionInterceptor transactionInterceptor;
+  public static final String APPLICATION_NAME_SPACE = "company-accounts.api.ch.gov.uk";
```
Comment (cburns93) [inline src/main/java/uk/gov/companieshouse/api/accounts/CompanyAccountsApplication.java]: these lines have been reindented to 2 spaces?

File: src/main/java/uk/gov/companieshouse/api/accounts/exception/handler/GlobalExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+
+  private void logError(Exception ex, ExceptionMessage exceptionMessage) {
+    HashMap<String, Object> message = new HashMap<>();
+    message.put("message", ex.getMessage());
+    message.put("error", exceptionMessage.getError());
+    LOGGER.error(exceptionMessage.getMessage(), ex, message);
+  }
+
+  private ExceptionMessage getExceptionMessage(Exception ex) {
```
Comment (cburns93) [inline src/main/java/uk/gov/companieshouse/api/accounts/exception/handler/GlobalExceptionHandler.java]: we have a potential worry in future of this if as well as the case statement being very large. Can we do this another way?

File: src/main/java/uk/gov/companieshouse/api/accounts/exception/handler/GlobalExceptionHandler.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+      RuntimeException.class, NoSuchAlgorithmException.class, Exception.class})
+  @ResponseStatus(value = HttpStatus.INTERNAL_SERVER_ERROR)
+  protected void handleException(Exception ex) {
+    logError(ex, getExceptionMessage(ex));
+  }
+
+  @ExceptionHandler(value = {NoHandlerFoundException.class})
+  @ResponseStatus(value = HttpStatus.NOT_FOUND)
+  protected void handleNoHandlerFoundException(Exception ex) {
+    logError(ex, ExceptionMessage.NO_HANDLER_FOUND_EXCEPTION);
```
Comment (jsimons1989) [inline src/main/java/uk/gov/companieshouse/api/accounts/exception/handler/GlobalExceptionHandler.java]: Does Spring run these in order that they are defined? If so, couldn't this be simplified to:
```java
@ExceptionHandler(value = {NoHandlerFoundException.class})
@ResponseStatus(value = HttpStatus.NOT_FOUND)
protected void handleNotFoundException(Exception ex) {...}

@ExceptionHandler(value = {ApiException.class})
protected void handleApiException(ApiException apiex) {...}

@ExceptionHandler(value = {Exception.class})
@ResponseStatus(value = HttpStatus.INTERNAL_SERVER_ERROR)
protected void handleException(Exception ex) {..}
```

File: src/main/java/uk/gov/companieshouse/api/accounts/exception/handler/GlobalExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    } else if (ex instanceof RuntimeException) {
+      return ExceptionMessage.RUN_TIME_EXCEPTION;
+    }
+    return ExceptionMessage.EXCEPTION;
+  }
+
+  private HttpStatus getHttpStatus(ApiException apiEx) {
+    switch (apiEx.exceptionMessage) {
+      case ACCOUNT_LINK_PRESENT:
+        return HttpStatus.CONFLICT;
```
Comment (jsimons1989) [inline src/main/java/uk/gov/companieshouse/api/accounts/exception/handler/GlobalExceptionHandler.java]: Why is an API exception being thrown in the code expecting this handler to pick up the exception and convert to the correct HTTP status? This is more likely to lead to bugs. The correct HTTP status should be returned from the controller where the error condition occurs.

**Comment by gpatrac** [comment]
Build Failed as Jenkins couldn't reach SONAR.

[INFO] --- sonar-maven-plugin:3.4.0.905:sonar (default-cli) @ company-accounts.api.ch.gov.uk ---
[INFO] User cache: /home/ci/.sonar/cache
[ERROR] SonarQube server [http://code-analysis.aws.chdev.org:9000] can not be reached

**Comment by gpatrac** [comment]
http://ch-jenkins-2k8.internal.ch:8080/jenkins/job/Build/job/chs/job/company-accounts.api.ch.gov.uk/job/feature/196/console - Green

File: src/main/java/uk/gov/companieshouse/api/accounts/exception/handler/GlobalExceptionHandler.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  @ResponseStatus(value = HttpStatus.INTERNAL_SERVER_ERROR)
+  protected void handleException(Exception ex) {
+    logError(ex, getExceptionMessage(ex));
+  }
+
+  private void logError(Exception ex, ExceptionMessage exceptionMessage) {
+    HashMap<String, Object> message = new HashMap<>();
+    message.put("message", ex.getMessage());
+    message.put("error", exceptionMessage.getMessage());
+    LOGGER.error(exceptionMessage.getError(), ex, message);
```
Comment (jsimons1989) [inline src/main/java/uk/gov/companieshouse/api/accounts/exception/handler/GlobalExceptionHandler.java]: From looking at the text in ExceptionMessage. Can this be simplified?
```java
HashMap<String, Object> message = new HashMap<>();
message.put("error", e.getClass());
LOGGER.error(e, message);
```

"message" will get set to e.getMessage() in the structured logger.

File: src/main/java/uk/gov/companieshouse/api/accounts/exception/handler/GlobalExceptionHandler.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  @ExceptionHandler(value = {NoHandlerFoundException.class})
+  @ResponseStatus(value = HttpStatus.NOT_FOUND)
+  protected void handleNoHandlerFoundException(Exception ex) {
+    logError(ex, ExceptionMessage.NO_HANDLER_FOUND_EXCEPTION);
+  }
+
+  @ExceptionHandler(value = {DataAccessException.class, IOException.class,
+      IllegalArgumentException.class, IllegalStateException.class, NullPointerException.class,
+      NoSuchAlgorithmException.class, RuntimeException.class, Exception.class})
+  @ResponseStatus(value = HttpStatus.INTERNAL_SERVER_ERROR)
```
Comment (jsimons1989) [inline src/main/java/uk/gov/companieshouse/api/accounts/exception/handler/GlobalExceptionHandler.java]: Wouldn't all these exceptions get recognised using the super exception types?
```java
@ExceptionHandler(value = {RuntimeException.class, Exception.class})
```

Comment (gpatrac) [reply src/main/java/uk/gov/companieshouse/api/accounts/exception/handler/GlobalExceptionHandler.java]: Yes, but they were there as we needed to log specific exception message, copied from Abridged. The ExceptionMessage enum is not required at all.

**Pull request merged**